File: test/util/test_util_runfiles.cc
// Copyright 2019 The gVisor Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <iostream>
#include <string>

#include "test/util/fs_util.h"
#include "test/util/test_util.h"
#include "tools/cpp/runfiles/runfiles.h"

namespace gvisor {
namespace testing {

std::string RunfilePath(std::string path) {
  static const bazel::tools::cpp::runfiles::Runfiles* const runfiles = [] {
    std::string error;
    auto* runfiles =
        bazel::tools::cpp::runfiles::Runfiles::CreateForTest(&error);
    if (runfiles == nullptr) {
      std::cerr << "Unable to find runfiles: " << error << std::endl;
    }
    return runfiles;
  }();

  if (!runfiles) {
    // Can't find runfiles? This probably won't work, but __main__/path is our
    // best guess.
    return JoinPath("__main__", path);
  }

  // Try to resolve the path as it was passed to us, and check that it exists
  // before returning.
  std::string runfile_path = runfiles->Rlocation(JoinPath("__main__", path));
  struct stat st = {};
  if (!runfile_path.empty() && stat(runfile_path.c_str(), &st) == 0) {
    // Found it.
    return runfile_path;
  }

  // You are not gonna like this, but go_binary data dependencies have an extra
  // directory name with a "_" suffix, so we must check for that path too.
  //
  // For example, a go_binary with name"//foo/bar:baz" will be placed in
  // "<runfiles_dir>/foo/bar/baz_/baz".
  //
  // See
  // https://github.com/bazelbuild/rules_go/blob/d2a3cf2d6b18f5be19adccc6a6806e0c3b8c410b/go/private/context.bzl#L137.
  absl::string_view dirname = Dirname(path);
  absl::string_view basename = Basename(path);
  std::string go_binary_path =
      JoinPath(dirname, absl::StrCat(basename, "_"), basename);
  return runfiles->Rlocation(JoinPath("__main__", go_binary_path));
}

}  // namespace testing
}  // namespace gvisor

File: pkg/test/testutil/testutil_runfiles.go
// Copyright 2018 The gVisor Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//go:build go1.1
// +build go1.1

package testutil

import (
	"fmt"
	"os"
	"path/filepath"
)

// FindFile searchs for a file inside the test run environment. It returns the
// full path to the file. It fails if none or more than one file is found.
func FindFile(path string) (string, error) {
	wd, err := os.Getwd()
	if err != nil {
		return "", err
	}

	// The test root is demarcated by a path element called "__main__". Search for
	// it backwards from the working directory.
	root := wd
	for {
		dir, name := filepath.Split(root)
		if name == "__main__" {
			break
		}
		if len(dir) == 0 {
			return "", fmt.Errorf("directory __main__ not found in %q", wd)
		}
		// Remove ending slash to loop around.
		root = dir[:len(dir)-1]
	}

	// Annoyingly, bazel adds the build type to the directory path for go
	// binaries, but not for c++ binaries. We use two different patterns to
	// to find our file.
	patterns := []string{
		// Try the obvious path first.
		filepath.Join(root, path),
		// If it was a go binary, use a wildcard to match the build
		// type. The pattern is: /test-path/__main__/directories/*/file.
		filepath.Join(root, filepath.Dir(path), "*", filepath.Base(path)),
	}

	for _, p := range patterns {
		matches, err := filepath.Glob(p)
		if err != nil {
			// "The only possible returned error is ErrBadPattern,
			// when pattern is malformed." -godoc
			return "", fmt.Errorf("error globbing %q: %v", p, err)
		}
		switch len(matches) {
		case 0:
			// Try the next pattern.
		case 1:
			// We found it.
			return matches[0], nil
		default:
			return "", fmt.Errorf("more than one match found for %q: %s", path, matches)
		}
	}
	return "", fmt.Errorf("file %q not found", path)
}
